core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -268,7 +290,9 @@ private boolean isEnoughVersionBasedAllocationsFound(IndexMetaData indexMetaData
     /**
      * Split the list of node shard states into groups yes/no/throttle based on allocation deciders
      */
-    private NodesToAllocate buildNodesToAllocate(ShardRouting shard, RoutingAllocation allocation, List<NodeGatewayStartedShards> nodeShardStates) {
+    private NodesToAllocate buildNodesToAllocate(RoutingAllocation allocation,
+                                                 List<NodeGatewayStartedShards> nodeShardStates,
+                                                 Function<RoutingNode, Decision> decisionFunction) {
```

Comment:
we could alternatively just pass a boolean `forceAllocate` instead of the decisionFunction.